nuxeo-features/nuxeo-automation/nuxeo-automation-features/src/main/java/org/nuxeo/ecm/automation/core/operations/services/FileManagerImport.java

```diff
@@ -73,13 +76,20 @@ protected DocumentModel getCurrentDocument() throws OperationException {
 
     @OperationMethod
     public DocumentModel run(Blob blob) throws OperationException, IOException {
+        // do the import in a long-running transaction to avoid timeouts
+        TransactionHelper.commitOrRollbackTransaction();
+        TransactionHelper.startTransaction(IMPORT_TX_TIMEOUT_SEC);
         DocumentModel currentDocument = getCurrentDocument();
         String path = currentDocument.getPathAsString();
         FileImporterContext fileCreationContext = FileImporterContext.builder(session, blob, path)
                                                                      .overwrite(overwite)
                                                                      .mimeTypeCheck(!noMimeTypeCheck)
                                                                      .build();
-        return fileManager.createOrUpdateDocument(fileCreationContext);
+        DocumentModel doc = fileManager.createOrUpdateDocument(fileCreationContext);
+        // go back to default transaction timeout
+        TransactionHelper.commitOrRollbackTransaction();
+        TransactionHelper.startTransaction();
```

Comment:
Isn't this the job of `TransactionHelperstartTransaction(int timeout)`?
See https://github.com/nuxeo/nuxeo/blob/master/nuxeo-runtime/nuxeo-runtime-jtajca/src/main/java/org/nuxeo/runtime/transaction/TransactionHelper.java#L368.

Reply:
No the `finally` block there is to set back the thread-local default transaction timeout. We have to do it this way because the `TransactionManager` API is badly designed, and you can't start a transaction with a given timeout, you have to set a default transaction timeout (in a thread-local) first.